CRM/Core/BAO/EntityTag.php

```diff
@@ -483,4 +483,37 @@ public static function buildOptions($fieldName, $context = NULL, $props = []) {
     return $options;
   }
 
+  /**
+   * This function deletes entity tags when a related entity is called.
+   *
+   * It is registered as a listener in \Civi\Core\Container::createEventDispatcher
+   *
+   * Ideally we will later nuance it so it's only called on delete....
+   *
+   * @param \Civi\Core\Event\PreEvent $event
+   */
+  public static function preDeleteOtherEntity($event) {
+    if ($event->action !== 'delete'
+      // Activity can call the pre hook for delete with no ID - this seems to be isolated to activity....
+      // @todo - what is the correct way to standardise?
+      || ($event->entity === 'Activity' && !$event->id)) {
+      return;
+    }
+    // This is probably fairly mild in terms of helping performance - a case could be made to check if tags
+    // exist before deleting (further down) as delete is a locking action.
+    if (!isset(Civi::$statics[__CLASS__]['tagged_entities'][$event->entity])) {
+      $entity = in_array($event->entity, ['Individual', 'Household', 'Organization']) ? 'Contact' : $event->entity;
+      $tableName = CRM_Core_DAO_AllCoreTables::getTableForEntityName($entity);
+      $used_for = CRM_Core_OptionGroup::values('tag_used_for');
+      Civi::$statics[__CLASS__]['tagged_entities'][$event->entity] = !empty($used_for[$tableName]) ? $tableName : FALSE;
+    }
+
+    if (Civi::$statics[__CLASS__]['tagged_entities'][$event->entity]) {
+      CRM_Core_DAO::executeQuery('DELETE FROM civicrm_entity_tag WHERE entity_table = %1 AND entity_id = %2',
+        [1 => [Civi::$statics[__CLASS__]['tagged_entities'][$event->entity], 'String'], 2 => [$event->id, 'Integer']]
+      );
```

Comment:
Firing off an ad-hoc query feels icky somehow. Shouldn't we call pre/post hooks before deleting entityTags? And shouldn't there already be a bao function to delete entityTags? And can we avoid an infinite loop in the process of calling a pre hook from within a pre hook listener?

Reply:
good point about the loop.

My instinct here was that this is kind of a bottom-of-the-food-chain entity  &  the efficiency of a single query makes sense

Reply:
Oh - the existing entity-tag functions hurt my head